src/common/grpc.rs

```diff
@@ -1067,6 +1069,49 @@ impl GrpcError {
     }
 }
 
+#[derive(Debug, Error)]
+pub struct ServerError {
+    grpc_error: GrpcError,
+    #[source]
+    cause: Option<anyhow::Error>,
+}
+
+impl ServerError {
+    pub fn new(grpc_error: GrpcError, cause: Option<anyhow::Error>) -> Self {
+        Self { grpc_error, cause }
+    }
+
+    pub fn to_status(&self) -> Status {
+        self.grpc_error.to_status(self.to_string())
+    }
+
+    pub fn status_code(&self) -> i32 {
+        self.grpc_error as i32
+    }
+}
+
+impl From<GrpcError> for ServerError {
+    fn from(grpc_error: GrpcError) -> Self {
+        Self {
+            grpc_error,
+            cause: None,
+        }
+    }
+}
+
+impl fmt::Display for ServerError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.cause {
+            Some(err) => write!(f, "{}: {}", self.grpc_error, err),
```

Comment:
you may be able to delegate to Error's Display implementation

Reply:
it was a lot more complicated to try to do this because the Debug and Display traits both have the `fmt` function with the same signature